### PR TITLE
Issue #251: ci:build patch-filter jar buring a build

### PR DESCRIPTION
**Pull request by HuGanghui**
Issue #251: ci:build patch-filter jar buring a build

this refers to https://github.com/checkstyle/eclipse-cs/pull/246

### CODE REVIEW

**Comment by romani** [comment]
please rebase and resolve CI failure

**Comment by HuGanghui** [comment]
@romani an error happened, I do not know how to solve it, please give me some help.

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (checkstyle-check) on project net.sf.eclipsecs.parent: Failed during checkstyle configuration: cannot initialize module TreeWalker - cannot initialize module AvoidDoubleBraceInitialization - Unable to instantiate 'AvoidDoubleBraceInitialization' class, it is also not possible to instantiate it as com.github.sevntu.checkstyle.checks.annotation.AvoidDoubleBraceInitialization, com.github.sevntu.checkstyle.checks.coding.AvoidDoubleBraceInitialization, com.github.sevntu.checkstyle.checks.design.AvoidDoubleBraceInitialization, com.github.sevntu.checkstyle.checks.naming.AvoidDoubleBraceInitialization, com.github.sevntu.checkstyle.checks.sizes.AvoidDoubleBraceInitialization, com.github.sevntu.checkstyle.checks.AvoidDoubleBraceInitialization, com.github.sevntu.checkstyle.AvoidDoubleBraceInitialization, AvoidDoubleBraceInitializationCheck, com.github.sevntu.checkstyle.checks.annotation.AvoidDoubleBraceInitializationCheck, com.github.sevntu.checkstyle.checks.coding.AvoidDoubleBraceInitializationCheck, com.github.sevntu.checkstyle.checks.design.AvoidDoubleBraceInitializationCheck, com.github.sevntu.checkstyle.checks.naming.AvoidDoubleBraceInitializationCheck, com.github.sevntu.checkstyle.checks.sizes.AvoidDoubleBraceInitializationCheck, com.github.sevntu.checkstyle.checks.AvoidDoubleBraceInitializationCheck, com.github.sevntu.checkstyle.AvoidDoubleBraceInitializationCheck. Please recheck that class name is specified as canonical name or read how to configure short name usage https://checkstyle.org/config.html#Packages. Please also recheck that provided ClassLoader to Checker is configured correctly. -> [Help 1]

```

File: pom.xml [hunk 1]
```diff
@@ -23,6 +23,12 @@
         <tycho-version>1.6.0</tycho-version>
         <tycho.scmUrl>scm:git:git://github.com:checkstyle/eclipse-cs.git</tycho.scmUrl>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <maven.checkstyle.plugin.version>3.1.0</maven.checkstyle.plugin.version>
+        <maven.sevntu.checkstyle.plugin.version>1.37.0</maven.sevntu.checkstyle.plugin.version>
+        <!-- it should be a version of checkstyle that is compatible/compiled with sevntu -->
+        <maven.sevntu.checkstyle.plugin.checkstyle.version>
+            8.29
```
Comment (romani) [inline pom.xml:30]: https://checkstyle.org/releasenotes.html#Release_8.30

  Problematic Check is introduced in next version.
Please use there 8.35, latest released.

**Comment by romani** [comment]
Please use https://github.com/checkstyle/eclipse-cs/blob/master/net.sf.eclipsecs.core/src/net/sf/eclipsecs/core/CheckstylePlugin.java#L1 for header file in checkstyle config

File: pom.xml (hunk 1 above)
Comment (romani) [reply pom.xml:30]: please see how I did it at https://github.com/checkstyle/sonar-checkstyle/pull/316

Comment (HuGanghui) [reply pom.xml]: done

**Comment by HuGanghui** [comment]
@romani here is a strange error in openjdk13, two others is ok now.

image

**Comment by HuGanghui** [comment]
and when I run `mvn clean verify` in my local, then a error will happen but it does not appear in CI:

The method disconnect(IPath, IProgressMonitor) from the type IFileBufferManager is deprecated
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] eclipse-cs Maven Parent 8.35.0-SNAPSHOT ............ SUCCESS [  4.642 s]
[INFO] eclipse-cs Target Definition ....................... SUCCESS [  1.985 s]
[INFO] eclipse-cs Branding Plugin ......................... SUCCESS [  2.114 s]
[INFO] Checkstyle Core Library Plugin ..................... SUCCESS [ 14.232 s]
[INFO] eclipse-cs Core Plugin ............................. FAILURE [  1.823 s]
[INFO] eclipse-cs Documentation Plugin .................... SKIPPED
[INFO] eclipse-cs UI Plugin ............................... SKIPPED
[INFO] eclipse-cs Sample Plugin ........................... SKIPPED
[INFO] eclipse-cs Feature ................................. SKIPPED
[INFO] eclipse-cs Update Site 8.35.0-SNAPSHOT ............. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 46.837 s
[INFO] Finished at: 2020-08-24T20:28:34+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:1.6.0:compile (default-compile) on project net.sf.eclipsecs.core: Compilation failure: Compilation failure: 
[ERROR] /Users/hgh/Documents/GitHub/checkstyle-contribution/checkstyle-tester/repositories/eclipse-cs/net.sf.eclipsecs.core/src/net/sf/eclipsecs/core/config/meta/MetadataFactory.java:[64] 
[ERROR]         import org.example.ModuleDetails;
[ERROR]                ^^^^^^^^^^^
[ERROR] The import org.example cannot be resolved
[ERROR] /Users/hgh/Documents/GitHub/checkstyle-contribution/checkstyle-tester/repositories/eclipse-cs/net.sf.eclipsecs.core/src/net/sf/eclipsecs/core/config/meta/MetadataFactory.java:[65] 
[ERROR]         import org.example.ModulePropertyDetails;
[ERROR]                ^^^^^^^^^^^
[ERROR] The import org.example cannot be resolved
[ERROR] /Users/hgh/Documents/GitHub/checkstyle-contribution/checkstyle-tester/repositories/eclipse-cs/net.sf.eclipsecs.core/src/net/sf/eclipsecs/core/config/meta/MetadataFactory.java:[66] 
[ERROR]         import org.example.XMLMetaReader;
[ERROR]                ^^^^^^^^^^^
[ERROR] The import org.example cannot be resolved
[ERROR] /Users/hgh/Documents/GitHub/checkstyle-contribution/checkstyle-tester/repositories/eclipse-cs/net.sf.eclipsecs.core/src/net/sf/eclipsecs/core/config/meta/MetadataFactory.java:[108] 
[ERROR]         private static Map<String, ModuleDetails> sModuleDetailsRepo;
[ERROR]                                    ^^^^^^^^^^^^^

**Comment by romani** [comment]
please do same as CI do:
`CMD="./build-metafile-gen.sh && ./build-patch-filter.sh && mvn install`

File: target/checkstyle-cachefile [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#Mon Aug 24 20:28:14 CST 2020
+module-resource*?\:config/sevntu_suppressions.xml=F0847992EF5169789F9CCDE1E0312EBA8C027703
+module-resource*?\:show.patch=5AB37DF7A2B1C28DA872C57C26358B2A8C4DA61C
+configuration*?=52B10C8F03C3D56953783503957FB08EAF5DB67B
```
Comment (romani) [inline target/checkstyle-cachefile:4]: add `target` folder to `.gitignore` 
such file should not be in GIT

File: build-patch-filter.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+set -e
+
+git diff HEAD~1 HEAD > show.patch
```
Comment (romani) [inline build-patch-filter.sh:4]: `git diff HEAD~1 HEAD > target/show.patch`

to not create any extra files in git repo and keep `git diff` empty.

File: target/checkstyle-cachefile (hunk 2 above)
Comment (HuGanghui) [reply target/checkstyle-cachefile:4]: done

File: build-patch-filter.sh (hunk 3 above)
Comment (HuGanghui) [reply build-patch-filter.sh]: done

**Comment by HuGanghui** [comment]
@romani 

>...please execute your config in eclipse-cs on multiple commits to make sure it works and could  produce some violations
Your report generator might come handy

if what you mean is that using report generator in ci, then it is not easy to use, because it need Checkstyle repository and Contribution repository, it just use report-generator in eclipse-cs, here is the result: https://github.com/checkstyle/patch-filters/issues/297

**Comment by romani** [comment]
>if what you mean is that using report generator in ci

Not in CI. Do on local and share report.

File: .travis.yml [hunk 4]
```diff
@@ -27,19 +27,19 @@ matrix:
...
         - DESC="install (openjdk8)"
-        - CMD="./build-metafile-gen.sh && mvn install && git diff"
+        - CMD="./build-metafile-gen.sh && ./build-patch-filter.sh && mvn install && git diff HEAD~2 HEAD~1 > target/show.patch  && mvn clean verify && git diff"
 
     # JDK 11 (Long Term Support version)
     - jdk: openjdk11
       env:
         - DESC="install (openjdk11)"
-        - CMD="./build-metafile-gen.sh && mvn install && git diff"
+        - CMD="./build-metafile-gen.sh && ./build-patch-filter.sh && mvn install && git diff HEAD~2 HEAD~1 > target/show.patch  && mvn clean verify && git diff"
```
Comment (romani) [inline .travis.yml:36]: remove `&& git diff HEAD~2 HEAD~1 > target/show.patch  && mvn clean verify`

`install` already to `verify`

File: build-patch-filter.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+set -e
+
+mkdir target
+git diff HEAD~1 HEAD > target/show.patch
+rm -rf patch-filter
```
Comment (romani) [inline build-patch-filter.sh:6]: add empty line above this line to separate patch creation and build

File: .gitignore [hunk 6]
```diff
@@ -6,3 +6,5 @@ net.sf.eclipsecs.core/lib/metadata-gen-1.0-SNAPSHOT.jar
 *.iml
 *.prefs
 .checkstyle
+net.sf.eclipsecs.core/lib/patch-filter-1.0-SNAPSHOT.jar
+target/
```
Comment (romani) [inline .gitignore:10]: add `patch-filter` to this file.

File: config/checkstyle_sevntu_checks.xml [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+  </module>
+
+  <module name="com.puppycrawl.tools.checkstyle.filters.SuppressionPatchFilter">
+    <property name="file" value="show.patch"/>
+    <property name="strategy" value="patchedline" />
+  </module>
+
+  <module name="TreeWalker">
+    <module name="com.puppycrawl.tools.checkstyle.filters.SuppressionJavaPatchFilter">
+      <property name="file" value="show.patch"/>
```
Comment (romani) [inline config/checkstyle_sevntu_checks.xml:22]: `target/show.patch` , the same for above

File: .travis.yml (hunk 4 above)
Comment (HuGanghui) [reply .travis.yml:36]: done

File: build-patch-filter.sh (hunk 5 above)
Comment (HuGanghui) [reply build-patch-filter.sh:6]: done

File: .gitignore (hunk 6 above)
Comment (HuGanghui) [reply .gitignore:10]: done

File: config/checkstyle_sevntu_checks.xml (hunk 7 above)
Comment (HuGanghui) [reply config/checkstyle_sevntu_checks.xml:22]: done

**Comment by romani** [comment]
@Calixte , FYI.

We will remove HACK with `build.....sh` script ones I deploy patch filter to maven repo.

Id there will be any false positives from Checkstyle validation of code, we might disable it,  but for now this will help us to start following checkstyle config from now, all new code should follow checkstyle style.

**Pull request merged**